text/orderer-v3.md

```diff
@@ -0,0 +1,320 @@
+ - **Block replication service**: Server-side API that sends blocks upon request. 
+ - **Block synchronization utility**: Pulls blocks from other ordering service nodes and appends them to the local ledger. 
+ - **Communication service**: Authenticates client connections, receives messages from remote nodes and sends message to remote nodes.
+ - **Ledger subsystem**: Appends blocks to persistent storage and retrieves blocks by their index.
+ - **Channel participation service**: Manages local addition and removal of ledgers per channel.
+ - **Transaction filtering**: Enforces access control and validity checks on transactions both received from clients or from nodes.
+ - **Membership Service Provider subsystem**: Classifies identities into principals and drives policy authorization. 
+
+In this RFC, it is proposed that these building blocks be made consensus independent, and the building blocks that cannot be consensus independent will be 
+part of the consensus plugin that wraps around the consensus independent building blocks. 
+
+Therefore, the new organization of the components in the new ordering service node of a given type will look as follows:
+
+```
+         ____________________________________________
+        |                                            |
+        | chain management--------consensus instance |
+        |       |      |                             |
+        |       |      -----------consensus instance |
+        |       |                      |             |
+        |       |_______               |             |
+        |       ________|______________|____         |                         
+        |       |                          |         |
+        |       |Ordering service framework|         |                         
+        |       |           (B)            |         | 
+        |       |__________________________|         |    
+        |                                            |
+        |         ordering service node        (A)   |
+        |____________________________________________|
+```
+
+The ordering service framework (denoted **B**) refers to all building blocks above which are consensus independent.
+Wrapping around the components of the ordering service framework is a layer (denoted **A**) which is consensus specific 
+and includes the chain management, and the consensus instances (one per channel).
+
+
+
+# Connecting the pieces together
+[connecting]: #connecting-the-pieces-together
+
+Now, having common and consensus independent building blocks that perform various functions,
+an ordering service for a given consensus protocol can be implemented without changing Fabric codebase. 
+
+The developers who implement the ordering service node will just import the needed packages from the Fabric ordering service framework,
+and will implement a binary that services the [Atomic Broadcast gRPC API](https://github.com/hyperledger/fabric-protos/blob/main/orderer/ab.proto#L80-L86).
```

Comment:
Isn't the part that implements the client facing Atomic Broadcast API in the common building blocks rather than in the consensus-specific code? I assumed at least the Atomic Broadcast Deliver service would be implemented by the common Block Replication Service.

Also I think how to implement a binary deserves a whole section, answering questions like:
- What exactly is necessary to implement it?
- What does the consensus plugin look like (I assume it is not a Go plugin per se)?
- Is there a specific consensus interface in mind yet? (I realize it may change during implementation phase)
- What assumptions are made about the types of consensus libaries that can be used? (I assume not any BFT library could be used)
- Is it expected that some integration code would be required to tie the consensus library into the ordering node consensus interface?

Reply:
> Isn't the part that implements the client facing Atomic Broadcast API in the common building blocks rather than in the consensus-specific code? 

It doesn't matter if you can use a package that implements it. The orderer can be externally seen as a gRPC service for the atomic broadcast API. How you implement it, is up to you.


>  I assumed at least the Atomic Broadcast Deliver service would be implemented by the common Block Replication Service.

It already is, see the latest revision to the RFC.


> Also I think how to implement a binary deserves a whole section, answering questions like:
>  -   What exactly is necessary to implement it?
>  -  What does the consensus plugin look like (I assume it is not a Go plugin per se)?
>  -  Is there a specific consensus interface in mind yet? (I realize it may change during implementation phase)
>  -   What assumptions are made about the types of consensus libaries that can be used? (I assume not any BFT library could be used)
>  -   Is it expected that some integration code would be required to tie the consensus library into the ordering node consensus interface?


That's the thing - there is no "consensus plugin" in the new architecture because each orderer binary now only has a single type of consensus. Hence, there is no consensus interface anymore because no one is expected to invoke it. 
It is the other way around - the orderer is the consensus and it uses the non-consensus specific building blocks as it needs, according to its consensus implementation.

Reply:
I was expecting a lot more details in this section. Primarily, the interfaces that the non-consensus specific building blocks would expose and how the orderer is expected to consume these with an example - say, how a `raft` based orderer would consume those interfaces at different occasions for achieving the consensus - mainly covering the important functionalities such as transaction submission, reconfiguration, and block cutting - as the primary motivation of this RFC is for others to understand the functions provided by these services so they can build orderer using their preferred consensus algorithm.

Reply:
@manish-sethi I'm not sure explicit interfaces for building blocks should be in an RFC. I also don't want to bind us to a specific contract should we later discover we need to make changes. 

Furthermore, we currently have the existing subcomponents in our code base, and they have existing APIs. I do not expect them to change too much so I'm not sure why it's important to specify in an RFC that a block replication API gets as input a block number and returns a block, and a communication API receives a message and a destination, etc. 

As for transaction submission, reconfiguration and block cutting - these are all completely orthogonal to this RFC, since they deal with the consensus itself. Here, we deal with what components a consensus needs. 
It is up to the consumer of the Fabric ordering service framework to decide how transactions are submitted, how nodes are being re-configured and how transactions are cut into batches.